domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt

```diff
@@ -0,0 +1,174 @@
+package org.oppia.domain.audio
+
+import android.content.Context
+import android.media.MediaPlayer
+import android.net.Uri
+import androidx.annotation.VisibleForTesting
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import org.oppia.util.data.AsyncResult
+import org.oppia.util.threading.BackgroundDispatcher
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.locks.ReentrantLock
+import javax.inject.Inject
+import kotlin.concurrent.withLock
+
+/** Controller which provides audio playing capabilities.
+ * [initializeMediaPlayer] should be used to download a specific audio track.
+ * [releaseMediaPlayer] should be used to clean up the controller's resources.
+ * See documentation for both to understand how to use them correctly. */
+class AudioPlayerController @Inject constructor(
+  val context: Context, @BackgroundDispatcher private val backgroundDispatcher: CoroutineDispatcher
+) {
+
+  inner class AudioMutableLiveData : MutableLiveData<AsyncResult<PlayProgress>>() {
+    override fun onActive() {
+      super.onActive()
+      observerActive = true
+    }
+
+    override fun onInactive() {
+      super.onInactive()
+      observerActive = false
+      stopUpdatingSeekBar()
+    }
+  }
+
+  enum class PlayStatus { PREPARED, PLAYING, PAUSED, COMPLETED}
+  class PlayProgress(val type: PlayStatus, val position: Int, val duration: Int)
+  class MediaPlayerAlreadyInitException : Exception()
+  class MediaPlayerNotInitException : Exception()
+
+  private val mediaPlayer: MediaPlayer by lazy { MediaPlayer() }
+  private val playState = AudioMutableLiveData()
+  private var nextUpdateJob: Job? = null
+  private val seekBarLock = ReentrantLock()
+
+  private var prepared = false
+  private var observerActive = false
+  private var mediaPlayerActive = false
+
+  private val SEEKBAR_UPDATE_FREQUENCY = TimeUnit.SECONDS.toMillis(1)
+
+  /* Call function to begin loading a audio sourced specified by url
+  *  MediaPlayer must be inactive when calling this function */
+  fun initializeMediaPlayer(url: String) {
+    if (mediaPlayerActive) throw MediaPlayerAlreadyInitException()
+    mediaPlayerActive = true
+    setMediaPlayerListeners()
+    prepareDataSource(url)
+  }
+
+  /* Call function to change data source of media player
+  *  Puts media player in a preparing state */
+  fun changeDataSource(url: String) {
+    prepared = false
+    stopUpdatingSeekBar()
+    mediaPlayer.reset()
+    prepareDataSource(url)
+  }
+
+  private fun setMediaPlayerListeners() {
+    mediaPlayer.setOnCompletionListener {
```

Comment:
Are these callbacks called on the UI thread?

Reply:
The callbacks happen on the same thread that creates the controller